### PR TITLE
Take Home Task

### PR DESCRIPTION
**Pull request by raunaqsachdev**
Testing Strategy: 
- More unit tests. All UI Components, stores have unit tests to test all edge cases etc.
- Fewer integration tests: All create files can do with simpler integration tests
- E2E tests: Since we've separated concerns as much as possible, we can make do with fewest amount of E2E tests.

### CODE REVIEW

File: packages/react/src/routes.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    CommandCenter
+}: AppRoutesProps) => {
+    return (
+        <BrowserRouter>
+            <Routes>
+                <Route element={<>
+                    <ProjectProvider>
+                        <CommandCenter />
+                        <Outlet />
+                    </ProjectProvider>
```
Comment (raunaqsachdev) [inline packages/react/src/routes.tsx:23]: Wrapped the entire app in `ProjectProviderContext`

This was useful for accessing project context from within CommandCenter. This also renders CommandCenter on all routes within the app

File: packages/react/src/components/CommandCenter/create.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        useEffect(() => {
+            commandCenterStore.syncProjectContext(projectContext);
+        }, [projectContext]);
+
+        return <CommandCenterComponent
+            isOpen={commandCenterStore.isOpen}
+            onClose={handleClose}
+            CommandCenterSuggestions={CommandCenterSuggestions}
+            CommandCenterInput={CommandCenterInput}
+        />
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/create.tsx:136]: We use Dependency Injection to inject pre-build components into CommandCenterComponent. 
This means the CommandCenterComponent is only responsible for it's own rendering and CSS.

This eventually helps in: 
- Avoiding prop drilling for props that'd be needed downstream
- Unncessary re-renders due to props being passed to children.
- Separation of concerns. The CommandCenterInput can essentially be treated as an individual component which would be easier to change/manipulate when needed.
- Cleaner separation of UI layer and business logic layers.
- Each unit is individually test-able.

File: packages/react/src/components/CommandCenter/create.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            action: () => suggestionsStore.highlightNextSuggestion(),
+            keys: KEY_BINDINGS.ARROW_DOWN,
+            element: inputRef
+        });
+
+        useKeybinding({
+            action: () => suggestionsStore.highlightPreviousSuggestion(),
+            keys: KEY_BINDINGS.ARROW_UP,
+            element: inputRef
+        });
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/create.tsx:45]: The input here is responsible for holding on to the focus and thus gets the keyboard events added to it.

File: packages/react/src/components/CommandCenter/commandCenterStore.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import { IReactionDisposer, makeAutoObservable, reaction } from "mobx";
+import { InputStore } from "./InputStore";
+import { SuggestionsStore } from "./SuggestionsStore";
+import { Suggestion } from "@/constants/commandTypes";
+import { ProjectContextType } from "@/contexts/Project/useProjectContext";
+import { debounce } from "lodash";
+
+
+export class CommandCenterStore {
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/commandCenterStore.ts:9]: There's a lot of business logic in here that could have been abstracted further. (Couldn't because short on time)
For example: 
- We could have a separate controller for managing the reactions and linked dependencies between the 3 stores. That'd help in separating the concerns from all 3 stores.

File: packages/react/src/components/CommandCenter/SuggestionsStore.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        const nextIndex = (currentIndex + 1) % this.suggestions.length;
+        this._highlightedSuggestion = this.suggestions[nextIndex];
+    }
+
+    highlightPreviousSuggestion() {
+        if (!this.suggestions.length) return;
+        const currentIndex = this.suggestions.findIndex(s => s === this._highlightedSuggestion);
+        const previousIndex = (currentIndex - 1 + this.suggestions.length) % this.suggestions.length;
+        this._highlightedSuggestion = this.suggestions[previousIndex];
+    }
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/SuggestionsStore.ts:43]: We do expensive find operations each time, but it's deemed acceptable as we are expected to have a small number of suggestions.

File: packages/react/src/components/CommandCenter/Suggestions.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import { Suggestion } from '@/constants/commandTypes';
+
+type CommandCenterSuggestionsProps = {
+  suggestions: Suggestion[];
+  onSuggestionClick: (suggestion: Suggestion) => void;
+  highlightedSuggestionCommand?: Suggestion;
+  setHighlightedSuggestionCommand: (command: Suggestion) => void;
+}
+
+export const CommandCenterSuggestions = memo(({ suggestions, onSuggestionClick, highlightedSuggestionCommand, setHighlightedSuggestionCommand }: CommandCenterSuggestionsProps) => {
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/Suggestions.tsx:12]: Isolated UI component that can easily be unit tested

File: packages/react/src/components/CommandCenter/create.tsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import { Input } from "./Input";
+import { useKeybinding } from "@/hooks/keyboardEventHandler";
+import { KEY_BINDINGS } from "@/constants/keybindings";
+import { Suggestion } from "@/constants/commandTypes";
+import { ProjectContextType, useProjectContext } from "@/contexts/Project/useProjectContext";
+import { InputStore } from "./InputStore";
+import { CommandCenterStore } from "./CommandCenterStore";
+
+
+const createCommandCenterInput = ({ suggestionsStore, inputStore, onSuggestionSelect }: {
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/create.tsx:15]: Each create function can have vi-tests written for it which would essentially be the equivalent of integration tests. We'd be testing the integration of different parts that come together within a create function. Each create function accepts dependencies which make it easier to test in isolation.

File: packages/react/src/components/CommandCenter/InputStore.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import { makeAutoObservable } from "mobx";
+
+
+const DEFAULT_INPUT = {
+    placeholder: 'What would you like to do?',
+    label: 'Command',
+    value: ''
+};
+
+export class InputStore {
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/InputStore.ts:10]: Simple store for managing the input component. We use an external store instead of React's internal state as this allows us to use the value of the store externally.

File: packages/react/src/components/CommandCenter/Input.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import styles from './styles.module.css';
+
+type InputProps = {
+    value: string;
+    onChange: (value: string) => void;
+    placeholder: string;
+    label: string;
+};
+
+export const Input = forwardRef<HTMLInputElement, InputProps>(function Input({
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/Input.tsx:11]: Simple input component. 
This component has some internal state as well since it's the kind of state we don't really need to access this state externally.

Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/Input.tsx:11]: forwardRef to add keyboard event handlers to the input ref. 

We wan't to keep concerns separate and the keyboard event handlers in this case aren't really related to input component's logic thus it's better to use a forward ref here. 

If it were just `Enter`, we could have added the keybinding within the component and skipped the `forwardRef` part

File: packages/react/src/components/CommandCenter/CommandCenter.tsx [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+import styles from './styles.module.css';
+
+type CommandCenterProps = {
+  isOpen: boolean;
+  onClose: () => void;
+  CommandCenterInput: React.ComponentType;
+  CommandCenterSuggestions: React.ComponentType;
+};
+
+export const CommandCenter: React.FC<CommandCenterProps> = ({ isOpen, onClose, CommandCenterSuggestions, CommandCenterInput }) => {
```
Comment (raunaqsachdev) [inline packages/react/src/components/CommandCenter/CommandCenter.tsx:13]: Simple UI component, no complex logic here

File: packages/react/src/App.tsx [hunk 11]
```diff
@@ -1,42 +1,62 @@
...
+      return defaultSuggestions.filter(suggestion =>
+        suggestion.shouldShow(projectContext)
+      );
+    }
+
+    return defaultSuggestions.filter(suggestion =>
+      suggestion.shouldShow(projectContext) &&
+      suggestion.label.toLowerCase().includes(input.toLowerCase())
+    );
+  };
```
Comment (raunaqsachdev) [inline packages/react/src/App.tsx:44]: The fetch suggestions has been abstracted away and can be made more powerful with NLP, fuzzy search or whatever we want to do.
